python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py

```diff
@@ -444,8 +542,14 @@ def description(self) -> str:
         return "waiting on upstream data to be up to date"
 
     def evaluate_for_asset(self, context: RuleEvaluationContext) -> RuleEvaluationResults:
-        asset_partitions_by_waiting_on_asset_keys = defaultdict(set)
-        for candidate in context.candidates:
+        asset_partitions_by_evaluation_data = defaultdict(set)
+
+        # only need to evaluate net-new candidates and candidates whose parents have changed
+        candidates_to_evaluate = (
+            context.get_candidates_not_skipped_on_previous_tick()
```

Comment:
Those results from the previous tick won't be accurate if either:

a) the candidate wasn't skipped on the previous tick (meaning no skip evaluation data from the previous tick is relevant)

or

b) one or more of the candidate's parents have been updated since the previous tick (as this can invalidate the previously-calculated result)

Reply:
To clarify, if I understand correctly, the cursor trscks things:
1. For this individual skip rule, which asset partitions were skipped on the previous tick
2. Across all rules, which assets partitions were skipped on the previous tick

I wondering why we're using (2) instead of (1). My reasons for preferring (1):
- Feels easier to reason about. I only need to think about what happened in this rule evaluation, not what happened in all the other rule evaluations.
- Maybe we don't need to track the global skip set on the cursor at all?

Reply:
gah you're very right

Reply:
specifically, we do actually need some information that's global across rules, but we do not need to store this information in a separate object on the cursor as the asset evaluations contain all the information we need.

the reason we need this "rule-spanning" information is that we can only re-use data for an asset if it wasn't requested or discarded on the previous tick, and determining if a partition was requested or discarded requires knowledge of what all of the rules evaluated to.

Reply:
> the reason we need this "rule-spanning" information is that we can only re-use data for an asset if it wasn't requested or discarded on the previous tick

why exactly is that?

Reply:
I can imagine cases where some rules don't want this behavior, but for materialize_on_missing / materialize_on_parent_updated, this is the case. In particular, imagine the following sequence, for `A -> B`, where B just has "materialize_on_missing" and "skip_on_parent_missing", and A has no policy:

On tick 1, B notices that it is missing, but also that A is missing, meaning both rules fire, and no materialization is requested.

Things remain in this state (both rules firing, no asset requested) until A is manually executed. During this time, the materialize_on_missing rule is continually firing because we remember its value from the previous tick.

On the next tick, we re-evaluate the skip rule, find that it is now false, and request to materialize B (because the materialize rule is still firing).

On the next tick after that, we now need to know that we no longer should fire that materialize rule. The reason that we no longer want to fire it is that we requested it on the previous tick. In order to determine that it was requested, we need to either keep track of that in the cursor (previous solution), or re-derive its status from the asset evaluation (current solution).

Reply:
Gotcha. I see why, for `materialize_on_missing` and `materialize_on_parent_updated`, we need to check whether the asset partition was requested or discarded on a previous tick.

Do we need to do check that in the skip rules though?

Reply:
Side question: in `materialize_on_missing` and `materialize_on_parent_updated`, how _do_ we check whether the asset partition was requested or discarded on a previous tick?

Reply:
re: accessing global information in the skip rules, I went on a journey on this one and my conclusion is that it is useful to access this information, but that it's not _actually_ global. In an example (C depends on A and B, all partitioned, everything is fully materialized):

Tick 1: lots of partitions of A are updated, resulting in lots of materialize_on_parent_updated rules firing. However, all of the partitions of B are currently outdated (or basically any other skip reason causes none of the partitions of C to fire).

Tick 2: we remember all of those materialize reasons from the previous tick, and start feeding them through our skip rules. One of those skip rules is (e.g.) skip_on_parent_missing. If we only look at the previous evaluation data for that specific rule, we won't find anything, as none of the parent partitions are actually missing. However, we know that all of those materialize reasons existed the previous tick, and therefore the skip_on_parent_missing rule must have evaluated those partitions, and so the fact that there's no evaluation data implies that it evaluated to false on the previous tick.

In the updated code, we're really just looking at the set of asset partitions which had at least one materialize reason on the previous tick, as all of those would have ended up being evaluated by all skip rules on the previous tick.

I'm saying that this is not *really* global / cross-rule information, as the thing we're using it for is specific to our rule (which asset partitions were evaluated by this rule last tick), it just so happens that the way we calculate that value is by looking at other rules, and that this value is constant across all skip rules.

Reply:
> If we only look at the previous evaluation data for that specific rule, we won't find anything, as none of the parent partitions are actually missing.

Is my understanding correct that the issue here is that the data stored inside `latest_evaluation_by_asset_key` doesn't allow us to distinguish between these two cases?
- Situations where we evaluated the skip rule on asset partition X, and the partition wasn't skipped.
- Situations where we never evaluated the skip rule on asset partition X.

Reply:
That's right yep, although I'd phrase it more specifically along the lines of "there is no explicit metadata written when a skip rules decides not to skip an asset partition, and therefore we have to infer the fact that it decided not to skip a partition by inferring the set of partitions it must have evaluated"

Reply:
Got it.

>  I went on a journey on this one and my conclusion is that it is useful to access this information, but that it's not actually global.

I think future readers will find this difficult to decipher, and it seems very easy for global logic to sneak in.

If there's no straightforward / performant way to track this on a cursor, could we factor this out into a helper method? E.g.

```python
class RuleEvaluationContext(...):
    def was_evaluated_on_last_tick(partition_key: Optional[str]) -> bool:
        """
        <Explanation of the "However, we know that all of those materialize reasons existed the previous tick, and therefore the skip_on_parent_missing rule must have evaluated those partitions, and so the fact that there's no evaluation data implies that it evaluated to false on the previous tick" reasoning> 
        """
        ...
```

Then, in the future, if we can find a way to replace this with an implementation that relies on fewer assumptions, we won't need to re-reason about as many implications.

Reply:
@sryza currently this is handled by `get_candidates_not_evaluated_on_previous_tick`, which I think is in the shape you're looking for, but should be renamed something more specific to skip rules

Reply:
Ah ok oops yes that basically solves it. I think we should just find a way to make it clear (a couple words in the docstring would be fine) that this is referring to being evaluated in this specific rule.